Fix and speed up Product feature seed and tests

jrafanie opened the pull request:
We have several duplicated identifiers in the `db/fixtures/miq_product_features.yml` when they must be unique.  The seed method will create on the first instance and update on each subsequent instance and the last one in the file always wins.  Since only the last one survives, I chose to delete these "all but last duplicates" cc @h-kataria @dclarizio 

Added a missing test for "providing additional product features used by the UI plugins" (from #1680)

Changed the tests to not run the seed against the enormous miq_product_features.yml file except when we try to detect duplicate identifiers in changes to this file.

Tests before:
`Finished in 10.4 seconds`

Tests after:
`Finished in 3.25 seconds`

MiqProductFeature.seed is also a little faster:
Initial seed:            3 s -> 2.3 s
Subsequent seed:  0.8 s -> 0.2 s

Fryguy commented:
> Finally, I also removed the unused "feature" to seed product features from the non-existing db/fixtures/miq_product_features directory.

I believe this is for downstream productization.  Can you verify?

Fryguy commented on app/models/miq_product_feature.rb:
```diff
@@ -77,50 +75,59 @@ def self.seed
   def self.seed_features
-    idents_from_hash = []
-    self.seed_from_hash(YAML.load_file(FIXTURE_YAML), idents_from_hash)
+    seen = Hash.new([])
```
This feels bad.  This means that every key will share the same Array.

```ruby
seen = Hash.new([])
seen[:x] << "a"
seen[:y]
# => ["a"]
```

Fryguy commented on app/models/miq_product_feature.rb:
```diff
@@ -77,50 +75,59 @@ def self.seed
+      end
+
+    feature, status = seed_feature(hash, db_feature)
+    seen[status] = [] unless seen.key?(status)
```
But then you do this?  I don't think you need the initialization at all up above, except maybe to just say `seen = {}`

Fryguy replied:
Or, initialize it properly with `seen = Hash.new { |h, k| h[k] = [] }`

Then, you don't need this line at all.

jrafanie replied:
I'll try that, thanks!

Fryguy commented on app/models/miq_product_feature.rb:
```diff
@@ -77,50 +75,59 @@ def self.seed
+      if features
+        features.detect { |f| f.identifier == hash[:identifier] }
+      else
+        where(:identifier => hash[:identifier]).first
```
This is going to get really expensive as you recurse.

jrafanie replied:
Yes, I pass it in to eliminate this N+1. The only place we don't is for seed specific features in the tests.  Normal seeding will pass in all of the features up front instead of querying them here.

Fryguy commented on app/models/miq_product_feature.rb:
```diff
@@ -77,50 +75,59 @@ def self.seed
+
+    db_feature =
+      if features
+        features.detect { |f| f.identifier == hash[:identifier] }
```
You may want to `index_by(&:identifier)` beforehand, and then pass in the indexed features.

Fryguy commented on app/models/miq_product_feature.rb:
```diff
@@ -77,50 +75,59 @@ def self.seed
+    seen[status] << hash[:identifier]
+
+    children.each do |child|
+      self.seed_from_hash(child, seen, feature, features)
```
If you index the featuresm then here you should update with the new feature...then you can pass in the Hash and not need to have the else/where up above.

Fryguy commented:
If identifier is meant to be unique...let's put a unique validation on the model.  Then you don't need to self-check.  Seed will just straight-up fail if there are duplicates.

jrafanie commented:
> Finally, I also removed the unused "feature" to seed product features from the non-existing db/fixtures/miq_product_features directory.
> I believe this is for downstream productization. Can you verify?

I looked, I couldn't see anything besides an ae datastore... @h-kataria @dclarizio can you confirm?

jrafanie commented:
> If identifier is meant to be unique...let's put a unique validation on the model. Then you don't need to self-check. Seed will just straight-up fail if there are duplicates.

It's already there.  Seed will create or update based on the identifier.  It was looping through each identifier section of the yaml, fetch the row (N+1) matching the record, creating or updating it.

For example, tail of a log which no product features + seeding showing that we had duplicate identifiers in the yaml that were first created, then updated:
```
[----] I, [2015-09-30T12:52:04.120682 #32449:3fd8b5460208]  INFO -- : MIQ(MiqProductFeature.seed_from_hash) Creating product feature: Identifier: [ab_group_new], Name: [Add a new Button Group]
[----] I, [2015-09-30T12:52:05.439592 #32449:3fd8b5460208]  INFO -- : MIQ(MiqProductFeature.seed_from_hash) Updating product feature: Identifier: [ab_group_new], Name: [Add Button Group]
```

Fryguy commented:
>>> Finally, I also removed the unused "feature" to seed product features from the non-existing db/fixtures/miq_product_features directory.
>> I believe this is for downstream productization. Can you verify?
> I looked, I couldn't see anything besides an ae datastore... @h-kataria @dclarizio can you confirm?

The other thing it might be used for is the UI module plugin piece.

dclarizio commented:
> The other thing it might be used for is the UI module plugin piece.

@martinpovolny?

jrafanie commented:
@Fryguy I think I applied all of your suggestions

jrafanie commented:
> Finally, I also removed the unused "feature" to seed product features from the non-existing db/fixtures/miq_product_features directory. cc @Fryguy @gtanzillo @dclarizio Please let me know if this was wrong, I can't see this being used.

Ok, rebased, @martinpovolny any idea about the above directory that doesn't exist anywhere?  I'd like to remove that feature if we're not using it.  Thanks!

jrafanie commented:
@martinpovolny @dclarizio Can you review/merge this? Thanks!

h-kataria commented:
@jzigmund can you look over changes in db/fixtures/miq_product_features.yml, i think you are renaming a feature in your PR https://github.com/ManageIQ/manageiq/pull/3987 that's being deleted by this PR.

jrafanie commented:
@h-kataria @jzigmund @martinpovolny Can we get this reviewed?  Part of the benefit of the revised spec is that it will now fail if we accidentally introduce a new duplicate identifier.  Currently, we have duplicate identifiers we had no idea we had, nor if there are parts of the UI that needed the now deleted duplicates, and finally, it was costly both in tests and in seeding to create and then update the same duplicated identifier as only the last one in "wins".

jrafanie commented:
Bump, this really needs to get in so as we add new product features, the tests are verifying that the identifiers are unique.

jrafanie commented:
@h-kataria @martinpovolny @jzigmund @himdel @skateman Can someone please review the `Finally, I also removed the unused "feature" to seed product features from the non-existing db/fixtures/miq_product_features directory` part of this PR?   Do we ever seed product features anywhere using a yaml file that is not db/fixtures/product_features.yml?  That feature is being removed in this PR since it's unused AFAIK.

martinpovolny commented:
@jrafanie : The possibility to add product features is needed for the UI plugins to work. See https://github.com/ManageIQ/manageiq/pull/1680. We need to provide a different way for the plugins to add product features if we remove the `db/fixtures/miq_product_features directory`.

jrafanie commented:
Thanks @martinpovolny, I'll fix this PR to not kill that feature since it's in use.

martinpovolny commented:
@jrafanie : @Fryguy did not like the way I hacked this originally. So we might use the opportunity to address his concerns.

Fryguy commented:
@martinpovolny Yeah, we can revisit, but not for this performance PR.

jrafanie commented:
Ok @martinpovolny @Fryguy, I kept the UI plugins feature of loading additional product features and added a test.  I also sped up the tests some more.  See updated description above.

chessbyte commented:
@Fryguy @martinpovolny please review
